folding-schemes/src/frontend/circom/mod.rs

```diff
@@ -7,20 +7,100 @@ use ark_relations::r1cs::{ConstraintSynthesizer, ConstraintSystemRef, SynthesisE
 use ark_std::fmt::Debug;
 use num_bigint::BigInt;
 use std::path::PathBuf;
+use std::rc::Rc;
+use std::{fmt, usize};
 
 use crate::frontend::FCircuit;
 use crate::Error;
 
 pub mod utils;
 use utils::CircomWrapper;
 
+type ClourserPointer<F> = Rc<dyn Fn (usize, Vec<F>, Vec<F>) -> Result<Vec<F>, Error>>;
+
+
+#[derive(Clone)]
+struct CustomCode<F: PrimeField> {
+    func: ClourserPointer<F>,
+}
+
+impl<F: PrimeField> fmt::Debug for CustomCode<F> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "Function pointer: {:?}",
+            std::any::type_name::<fn(i32) -> i32>()
+        )
+    }
+}
+
 /// Define CircomFCircuit
 #[derive(Clone, Debug)]
 pub struct CircomFCircuit<F: PrimeField> {
     circom_wrapper: CircomWrapper<F>,
-    state_len: usize,
-    external_inputs_len: usize,
+    pub state_len: usize,
+    pub external_inputs_len: usize,
     r1cs: CircomR1CS<F>,
+    custom_code: Option<CustomCode<F>>,
+}
+
+impl<F: PrimeField> CircomFCircuit<F> {
+    pub fn set_custom_code(
```

Comment:
Could the `custom_code` references reference also to the `step_native`?
Something like
```suggestion
    pub fn set_custom_step_native(
```

And similar for the other `custom_code` references, like, in the `CircomFCircuit` struct,
`custom_code` -->`custom_step_native`
`struct CustomCode` --> `struct CustomStepNative`


The idea is to make it clear that is the custom code for the `step_native` method (not for the `generate_step_constraints` or other parts of the `code`).